File: backend/go/internal/capabilities/capabilities.go
package capabilities

import (
	"os"
	"strings"
)

// Name is a server-side product capability. Unknown names are always disabled.
type Name string

const (
	WorkspaceAuth                         Name = "workspace.auth"
	WorkspaceResume                       Name = "workspace.resume"
	WorkspacePublicImport                 Name = "workspace.public_import"
	WorkspaceATSAssistance                Name = "workspace.ats_assistance"
	WorkspaceKnowledgeHub                 Name = "workspace.knowledge_hub"
	WorkspaceInterviewPrep                Name = "workspace.interview_prep"
	WorkspaceApplicationTrack             Name = "workspace.application_tracker"
	AutonomousBrowser                     Name = "autonomous.browser"
	WorkspaceIsolatedComputer             Name = "workspace.isolated_computer"
	WorkspaceLocalBrowserBridge           Name = "workspace.local_browser_bridge"
	WorkspaceLocalBrowserSensitiveActions Name = "workspace.local_browser_sensitive_actions"
	WorkspaceComputerSubmission           Name = "workspace.computer_submission"
	AutonomousATSSubmit                   Name = "autonomous.ats_submit"
	AutonomousGmail                       Name = "autonomous.gmail"
	AutonomousMessaging                   Name = "autonomous.messaging"
	AutonomousBilling                     Name = "autonomous.billing"
	AutonomousIrreversible                Name = "autonomous.irreversible_jobs"
)

var known = map[Name]struct{}{
	WorkspaceAuth: {}, WorkspaceResume: {}, WorkspacePublicImport: {},
	WorkspaceATSAssistance: {}, WorkspaceKnowledgeHub: {}, WorkspaceInterviewPrep: {},
	WorkspaceApplicationTrack: {}, WorkspaceIsolatedComputer: {}, WorkspaceLocalBrowserBridge: {},
	WorkspaceLocalBrowserSensitiveActions: {}, WorkspaceComputerSubmission: {}, AutonomousBrowser: {}, AutonomousATSSubmit: {},
	AutonomousGmail: {}, AutonomousMessaging: {}, AutonomousBilling: {}, AutonomousIrreversible: {},
}

type Registry struct {
	enabled map[Name]bool
}

// NewFromEnv uses a conservative policy: workspace capabilities are enabled in
// development, but every capability is disabled in staging/production until an
// operator explicitly enables the workspace bundle. Autonomous capabilities are
// always disabled unless their individual flag is explicitly true.
func NewFromEnv() *Registry {
	env := strings.ToLower(strings.TrimSpace(os.Getenv("APP_ENV")))
	workspaceDefault := env != "production" && env != "prod" && env != "staging"
	registry := &Registry{enabled: make(map[Name]bool, len(known))}
	for name := range known {
		defaultValue := workspaceDefault && strings.HasPrefix(string(name), "workspace.")
		registry.enabled[name] = envBool(capabilityEnvKey(name), defaultValue)
	}
	return registry
}

func (r *Registry) Enabled(name Name) bool {
	if r == nil {
		return false
	}
	if _, ok := known[name]; !ok {
		return false
	}
	return r.enabled[name]
}

func (r *Registry) SetForTest(name Name, enabled bool) {
	if r == nil {
		return
	}
	if _, ok := known[name]; ok {
		r.enabled[name] = enabled
	}
}

func capabilityEnvKey(name Name) string {
	value := strings.ToUpper(strings.ReplaceAll(strings.ReplaceAll(string(name), ".", "_"), "-", "_"))
	return "CAPABILITY_" + value
}

func envBool(key string, fallback bool) bool {
	value, ok := os.LookupEnv(key)
	if !ok {
		return fallback
	}
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

File: backend/python/app/services/capabilities.py
"""Server-owned launch capability registry.

Unknown capabilities are disabled. Candidate-controlled workspace capabilities
are enabled only outside staging/production by default; high-risk capabilities
require their own explicit environment flag in every environment.
"""
from __future__ import annotations

import os
from enum import StrEnum


class Capability(StrEnum):
    WORKSPACE_AUTH = "workspace.auth"
    WORKSPACE_RESUME = "workspace.resume"
    WORKSPACE_PUBLIC_IMPORT = "workspace.public_import"
    WORKSPACE_ATS_ASSISTANCE = "workspace.ats_assistance"
    WORKSPACE_KNOWLEDGE_HUB = "workspace.knowledge_hub"
    WORKSPACE_INTERVIEW_PREP = "workspace.interview_prep"
    WORKSPACE_APPLICATION_TRACK = "workspace.application_tracker"
    WORKSPACE_EXTERNAL_RESEARCH = "workspace.external_research"
    WORKSPACE_EXTERNAL_RESEARCH_FIRECRAWL = "workspace.external_research.firecrawl"
    WORKSPACE_EXTERNAL_RESEARCH_APIFY = "workspace.external_research.apify"
    INTEGRATION_A2A_FEDERATION = "integration.a2a_federation"
    AUTONOMOUS_BROWSER = "autonomous.browser"
    WORKSPACE_ISOLATED_COMPUTER = "workspace.isolated_computer"
    WORKSPACE_LOCAL_BROWSER_BRIDGE = "workspace.local_browser_bridge"
    WORKSPACE_LOCAL_BROWSER_SENSITIVE_ACTIONS = "workspace.local_browser_sensitive_actions"
    WORKSPACE_COMPUTER_SUBMISSION = "workspace.computer_submission"
    AUTONOMOUS_ATS_SUBMIT = "autonomous.ats_submit"
    AUTONOMOUS_GMAIL = "autonomous.gmail"
    AUTONOMOUS_MESSAGING = "autonomous.messaging"
    AUTONOMOUS_BILLING = "autonomous.billing"
    AUTONOMOUS_IRREVERSIBLE = "autonomous.irreversible_jobs"


def _enabled(name: Capability) -> bool:
    environment = os.getenv("APP_ENV", "development").strip().lower()
    workspace_default = environment not in {"production", "prod", "staging"}
    default = workspace_default and name.value.startswith("workspace.")
    key = "CAPABILITY_" + name.value.upper().replace(".", "_").replace("-", "_")
    raw = os.getenv(key)
    if raw is None:
        return default
    return raw.strip().lower() in {"1", "true", "yes", "on"}


def capability_enabled(name: Capability | str) -> bool:
    try:
        capability = name if isinstance(name, Capability) else Capability(name)
    except ValueError:
        return False
    return _enabled(capability)


def require_capability(name: Capability | str) -> None:
    """Raise a stable machine-readable error for a disabled capability."""
    from fastapi import HTTPException

    if not capability_enabled(name):
        raise HTTPException(
            status_code=423,
            detail={
                "code": "disabled_by_launch_scope",
                "capability": str(name),
                "message": "This capability is not enabled for the current deployment scope.",
            },
        )
